lib/ansible/module_utils/network/nxos/config/telemetry/telemetry.py

```diff
@@ -0,0 +1,298 @@
+            saved_ids = []
+            for playvals in want[td['name']]:
+                valiate_input(playvals, td['name'], self._module)
+                if playvals['id'] in saved_ids:
+                    continue
+                saved_ids.append(playvals['id'])
+                resource_key = td['cmd'].format(playvals['id'])
+                # Only build the NxosCmdRef object for the subscription module parameters.
+                self._module.params['config'] = get_module_params_subsection(ALL_MP, td['type'], playvals['id'])
+                cmd_ref[td['type']]['ref'].append(NxosCmdRef(self._module, td['obj']))
+                ref = cmd_ref[td['type']]['ref'][-1]
+                ref.set_context([resource_key])
+                ref.get_existing(device_cache)
+                ref.get_playvals()
+
+        if state == 'overridden':
+            if want == have:
+                return []
+            commands = self._state_overridden(cmd_ref, want, have)
+        elif state == 'deleted':
+            commands = self._state_deleted(cmd_ref)
+        elif state == 'merged':
+            if want == have:
+                return []
+            commands = self._state_merged(cmd_ref)
+        elif state == 'replaced':
+            if want == have:
+                return []
+            commands = self._state_replaced(cmd_ref)
+        return commands
+
+    @staticmethod
+    def _state_overridden(cmd_ref, want, have):
+        """ The command generator when state is overridden
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        return commands
+
+    @staticmethod
+    def _state_replaced(cmd_ref):
+        """ The command generator when state is replaced
+        :rtype: A list
+        :returns: the commands necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        commands = []
+        return commands
+
+    @staticmethod
+    def _state_merged(cmd_ref):
+        """ The command generator when state is merged
+        :rtype: A list
+        :returns: the commands necessary to merge the provided into
+                  the current configuration
+        """
+        commands = cmd_ref['TMS_GLOBAL']['ref'][0].get_proposed()
+
+        if cmd_ref['TMS_DESTGROUP'].get('ref'):
+            for cr in cmd_ref['TMS_DESTGROUP']['ref']:
+                commands.extend(cr.get_proposed())
+
+        if cmd_ref['TMS_SENSORGROUP'].get('ref'):
+            for cr in cmd_ref['TMS_SENSORGROUP']['ref']:
+                commands.extend(cr.get_proposed())
+
+        if cmd_ref['TMS_SUBSCRIPTION'].get('ref'):
+            for cr in cmd_ref['TMS_SUBSCRIPTION']['ref']:
+                commands.extend(cr.get_proposed())
+
+        return remove_duplicate_context(commands)
+
+    @staticmethod
+    def _state_deleted(cmd_ref):
```

Comment:
@mikewiebe Just a question before this gets in.
`state_deleted` removes all the parameters that the resource has on the device which the module manages even when the parameters are not given in the task, right?

Reply:
No, it only deletes the parameters specified in the playbook.  State `replaced` would delete parameters not given in the task.  If deleted should also do that, it seem very counter intuitive.

Reply:
Also, if both delete and replaced remove all config not specified in the task, how can you only delete what you want?

Reply:
Here is a concrete example:

Let's say I have the following state on my device:

```yaml
- name: Setup - Configure Telemetry
  nxos_telemetry:
    state: 'merged'
    config:
      certificate:
        key: /bootflash/server.key
        hostname: localhost
      compression: gzip
      source_interface: "{{source_interface|default(omit)}}"
      vrf: management
      destination_groups:
        - id: 2
          destination:
            ip: 192.168.0.1
            port: 50001
            protocol: grpc
            encoding: gpb
        - { id: 2, destination: {ip: 192.168.0.2, port: 60001, protocol: gRPC, encoding: GPB}}
        - { id: 10, destination: {ip: 192.168.0.1, port: 50001, protocol: Grpc, encoding: gPB}}
        - { id: 10, destination: {ip: 192.168.0.2, port: 60001, protocol: gRPC, encoding: gpb}}
      sensor_groups:
        - { id: 8, data_source: NX-API, path: {name: sys/bgp, depth: 0, query_condition: foo, filter_condition: foo}}
        - { id: 2, data_source: NX-API, path: {name: sys/bgp/inst, depth: unbounded, query_condition: foo, filter_condition: foo}}
        - { id: 55, data_source: DME, path: {name: 'sys/bgp/inst/dom-default/peer-[10.10.10.11]/ent-[10.10.10.11]', depth: 0, query_condition: foo, filter_condition: foo}}
        - { id: 55, data_source: DME, path: {name: sys/ospf, depth: 0, query_condition: foo, filter_condition: 'or(eq(ethpmPhysIf.operSt,"down"),eq(ethpmPhysIf.operSt,"up"))'}}
    provider: "{{ connection }}"
```

I decide that I want to `delete` the source_interface, destination groups 2, 10.  It would not impact any other global settings, sensor_groups, destination_groups or subscriptions on the device.

```yaml
- name: Setup - Configure Telemetry
  nxos_telemetry:
    state: 'deleted'
      source_interface: "{{source_interface|default(omit)}}"
      destination_groups:
        - { id: 2}
        - { id: 10}
    provider: "{{ connection }}"
```

Reply:
cc @cidrblock @ikhan2010

Reply:
I had a chance to review this with @cidrblock.  I will update `state: deleted` to mean, purge all telemetry config.

Reply:
Thanks Mike!

Reply:
@trishnaguha Changes made.